float32 left1
float32 right1
float32 left2
float32 right2